src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -783,6 +785,372 @@ bool AdaptDispatch::SelectiveEraseInLine(const DispatchTypes::EraseType eraseTyp
     }
 }
 
+// Routine Description:
+// - Changes the attributes of each cell in a rectangular area of the buffer.
+// Arguments:
+// - textBuffer - Target buffer to be changed.
+// - changeRect - A rectangular area of the buffer that will be affected.
+// - changeOp - Function that will be applied to each of the attributes.
+// Return Value:
+// - <none>
+void AdaptDispatch::_ChangeRectAttributes(TextBuffer& textBuffer, const til::rect& changeRect, std::function<void(TextAttribute&)> changeOp)
```

Comment:
`std::function` is a very large object. I think it'd be better to pass it as `const&` here and down in `_ChangeRectOrStreamAttributes`.

---

I believe there's a way to make this work without injecting "dynamic" behavior. There's only 2 versions of `changeOp`. Wouldn't it be possible then to make a `struct ChangeOp` which contains the info for both, the change and inversion code? The benefit in my view would be that the control flow doesn't "jump" and can be easier to follow and (hopefully) to debug and that the `ChangeOp` contents can be easily inspected in a debugger.

Something like this (pseudocode):
```cpp
struct ChangeOp {
    CharacterAttributes resetAttrMask = CharacterAttributes::All;
    CharacterAttributes setAttrMask = CharacterAttributes::Normal;
    CharacterAttributes reverseMask = CharacterAttributes::Normal;
    std::optional<TextColor> foreground;
    std::optional<TextColor> background;
};

void AdaptDispatch::_ChangeRectAttributes(TextBuffer& textBuffer, const til::rect& changeRect, const ChangeOp& op)
{
            // ...
            for (auto col = changeRect.left; col < changeRect.right; col++)
            {
                auto attr = attrs.GetAttrByColumn(col);

                auto characterAttributes = attr.GetCharacterAttributes();
                characterAttributes &= op.resetAttrMask;
                characterAttributes |= op.setAttrMask;
                characterAttributes ^= op.reverseMask;
                attr.SetCharacterAttributes(characterAttributes);
                if (op.foreground)
                {
                    attr.SetForeground(*op.foreground);
                }
                if (op.background)
                {
                    attr.SetBackground(*op.background);
                }

                attrs.Replace(col, col + 1, attr);
            }
            // ...
}

bool AdaptDispatch::ChangeAttributesRectangularArea(const VTInt top, const VTInt left, const VTInt bottom, const VTInt right, const VTParameters attrs)
{
    // ...
    ChangeOp changeOp{
        .resetAttrMask = resetAttrMask,
        .setAttrMask = setAttrMask,
        .foreground = foregroundChanged ? std::optional{ foreground } : std::nullopt,
        .background = backgroundChanged ? std::optional{ background } : std::nullopt,
    };
    _ChangeRectOrStreamAttributes({ left, top, right, bottom }, changeOp);
    // ...
}

bool AdaptDispatch::ReverseAttributesRectangularArea(const VTInt top, const VTInt left, const VTInt bottom, const VTInt right, const VTParameters attrs)
{
        // ...
        ChangeOp changeOp{
            .reverseMask = reverseMask,
        };
        _ChangeRectOrStreamAttributes({ left, top, right, bottom }, changeOp);
        // ...
}
```

Reply:
Yeah. That seems like a good idea. It'll have to wait until the weekend, but I'll give it a try.